docs/man/fdconsole.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "FDCONSOLE" "1" "2016-12-11" "" ""
.
.SH "NAME"
\fBfdconsole\fR \- invoke FramerD interpeter
.
.SH "SYNOPSIS"
\fBfdconsole\fR [\fIconfig\fR|\fIfile\fR\.\.\.]
.
.SH "DESCRIPTION"
This command launches the FramerD console/terminal program, a read\-eval\-print loop which allows you to evaluate expressions and see their values\.
.
.P
The arguments to \fBfdconsole\fR are either filenames or \fIconfig\fR entries of the form \fB**setting**=*name*\fR\. The files are loaded and the config entries processed in order of appearance and config entries may specify multiple values for a setting\.
.
.P
When normally started, \fBfdconsole\fR uses the \fBeditline\fR library to provide simple command line editing and history functions\. This can be suppressed by the environment variables \fBNO_EDITLINE\fR or \fBINSIDE_EMACS\fR\.
.
.P
Some config entries may generate useful side\-effects defined by either the FramerD core or loaded modules\. Some useful built\-in config settings are:
.
.TP
\fBLOADFILE=*path*\fR
loads the file \fIpath\fR during startup\.
.
.TP
\fBMODULES=*module name*\fR
loads a FramerD module into the image and the default environment\.
.
.TP
\fBLOADPATH=*path*\fR
specifies a search path for modules, as a colon separated list of patterns where the \'%\' character is replaced by the module name\. For example \fB/opt/mods/%/module\.scm:/opt/mods/%:\fR causes the system to look for a module \fBfoo/x\fR in \fB/opt/mods/foo/x/module\.scm\fR and \fB/opt/mods/foo/x\.scm\fR\. Paths can be relative to the current directory by starting with \'\.\' and can also include ~home directory references\.
.
.TP
\fBLOGLEVEL=*level*\fR
specifies the default loglevel for notifications where level is either a positive integer or symbolic name (like \fBNOTICE\fR or \fBDEBUG\fR or \fBINFO\fR)\.
.
.SH "DIAGNOSTICS"
Returns 0 when exiting normally, 1 otherwise\. It also displays a short preamble on starting and reports its time and memory usage on exit\.
.
.SH "FILES"
On startup, \fBfdconsole\fR normally reads an \fB\.fdconfig\fR file from the user\'s home directory (if it exists)\. It can include \fBMODULES\fR or \fBLOADFILE\fR declarations\. The loading of the \fB\.fdconfig\fR file can be suppressed via the environment variable \fBFD_SKIP_DOTLOAD\fR or the config variable \fBDOTLOAD\fR\.
.
.SH "ENVIRONMENT"
As with most FramerD programs, config variables can be set through environment variables prefixed with \fBFD\fR, so \fBFD_LOGLEVEL\fR will set the default loglevel\.Ken Haase \fIkh@beingmeta\.com\fR
.
.SH "SEE ALSO"
fdexec(1), framerd_config(8)


debian/changelog
framerd (3.9.1-13) unstable; urgency=medium

  * Fix packaging descriptions
  * git patch version framerd-3.9.1-31-g2af7e18

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 11:11:23 -0400

framerd (3.9.1-12) unstable; urgency=medium

  * Split out framerd into framerd-core and framerd-builtin (which provides framerd-modules)
  * git patch version framerd-3.9.1-29-ge3a33fa

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 10:58:54 -0400

framerd (3.9.1-11) unstable; urgency=low

  * More adjustments to libmysqlclient dependencies
  * git patch version framerd-3.9.1-25-g880479a

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 10:12:25 +0000

framerd (3.9.1-10) unstable; urgency=low

  * Fix libmysqlclient dependency
  * Save alternate booksystem files in config directory
  * git patch version framerd-3.9.1-23-g4def5c8

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 09:17:44 +0000

framerd (3.9.1-9) unstable; urgency=low

  * Add libzip2 dependency to packaging
  * Fix FramerD systemd configuration
  * git patch version framerd-3.9.1-18

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 14:47:16 +0000

framerd (3.9.1-8) unstable; urgency=low

  * Add optional arg to empty-string? to treat non-breaking spaces as content
  * git patch version framerd-3.9.1-15-gff06aeb

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 03:51:54 +0000

framerd (3.9.1-7) unstable; urgency=low

  * Fix bug in optimize code
  * git patch version framerd-3.9.1-12-g99c9b27

 -- Ken Haase <kh@beingmeta.com>  Sat, 30 Apr 2016 22:51:01 +0000

framerd (3.9.1-6) unstable; urgency=medium

  * Repackage for alternate libmagickwand under xenial
  * git patch version framerd-3.9.1-9-gad2c5fc

 -- Ken Haase <kh@beingmeta.com>  Fri, 29 Apr 2016 13:16:58 -0400

framerd (3.9.1-5) unstable; urgency=medium

  * Improve error capture and handling including interrupts
  * Add fdstatus and fdlog commands
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Thu, 31 Mar 2016 07:01:55 -0400

framerd (3.9.0-4) unstable; urgency=medium

  * Fix build flags to remove needless dependencies
  * git patch version framerd-3.9-15-ge4f2745

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 17:35:42 -0400

framerd (3.9.0-3) unstable; urgency=low

  * Fix opcode handling bugs
  * git patch version framerd-3.9-6-gc9661e4

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 11:14:47 -0400

framerd (3.9.0-2) unstable; urgency=low

  * Fix git tag and executable heralds
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 21:13:27 -0400

framerd (3.9.0-1) unstable; urgency=low

  * Fix systemd scripts
  * Added fdstatus script
  * Fix pool-load to set error context
  * Fixes to opcode bugs
  * Bumped minor release number

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 16:55:24 -0400

framerd (3.8.4-63) unstable; urgency=low

  * Fix libbson/libmongoc dependencies
  * git patch verison framerd-3.8.4-177-gdb5c379

 -- Ken Haase <kh@beingmeta.com>  Sun, 27 Mar 2016 00:31:54 +0000

framerd (3.8.4-62) unstable; urgency=low

  * Fix tidy5 dependencies
  * git patch version framerd-3.8.4-176-g33eb981

 -- Ken Haase <kh@beingmeta.com>  Sat, 26 Mar 2016 14:25:11 +0000

framerd (3.8.4-61) unstable; urgency=low

  * Add smarter immediate type checking
  * Fix some opcode/optimize bugs
  * git patch version framerd-3.8.4-173-gdc93e81

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 21:33:36 -0400

framerd (3.8.4-60) unstable; urgency=low

  * Fixes to xmlparsing
  * Fix some applytest error cases (whoops)
  * Check irritant pointer when calling fd_err
  * Improve backtrace CSS
  * git patch version framerd-3.8.4-168-g8a0394e

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 18:21:37 -0400

framerd (3.8.4-59) unstable; urgency=low

  * Add handling of extpool lockfn
  * make DOM cleanup take more initiative
  * git patch version framerd-3.8.4-156-g43cb8ea

 -- Ken Haase <kh@beingmeta.com>  Sat, 19 Mar 2016 19:51:14 +0000

framerd (3.8.4-58) unstable; urgency=low

  * Add set-procedure-args!
  * Add functions for augmenting errors in ONERROR handlers
  * Fixes to optimize module
  * Wrap primitives execution in call contours to convert setmped errors
    to return errors
  * git patch version framerd-3.8.4-153-gcd07c91

 -- Ken Haase <kh@beingmeta.com>  Mon, 14 Mar 2016 18:26:35 -0400

framerd (3.8.4-57) unstable; urgency=low

  * Pass priority on to logger from logif+
  * git patch version framerd-3.8.4-147-gfb67432

 -- Ken Haase <kh@beingmeta.com>  Sat, 05 Mar 2016 22:12:07 -0500

framerd (3.8.4-56) unstable; urgency=low

  * Fix bug in JWT changes
  * git patch version framerd-3.8.4-145-gebeeed4

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 10:43:32 -0500

framerd (3.8.4-55) unstable; urgency=low

  * Add RSA algorithm to JWT
  * Improve and fix parts of JWT module
  * Add third arg (urish) to ->base64 functions to use - and _
    for + and /
  * git patch version framerd-3.8.4-143-gdb7869b

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 08:57:17 -0500

framerd (3.8.4-54) unstable; urgency=low

  * Record boot handler and status in FramerD run dir
  * git patch version framerd-3.8.4-132-g22819cc (pre commit)

 -- Ken Haase <kh@beingmeta.com>  Mon, 29 Feb 2016 23:08:03 +0000

framerd (3.8.4-53) unstable; urgency=low

  * Fixes to daemon/servlet management scripts
  * git patch version framerd-3.8.4-128

 -- Ken Haase <kh@beingmeta.com>  Sun, 28 Feb 2016 20:09:55 +0000

framerd (3.8.4-52) unstable; urgency=low

  * Fix set-cookie to not automatically add dot to cookie domains
  * Fixes and refinements to jwt and jwt/auth
  * Make parsetime handle non-string arguments
  * git patch version framerd-3.8.4-126

 -- Ken Haase <kh@beingmeta.com>  Fri, 26 Feb 2016 22:58:38 +0000

framerd (3.8.4-51) unstable; urgency=low

  * Remove uridecode handling of '+' escapes
  * git patch version framer-3.8.4-121

 -- Ken Haase <kh@beingmeta.com>  Tue, 23 Feb 2016 18:51:41 +0000


etc/fdconsole.el
;;; fdconsole.el --- emacs mode for the FramerD console   -*- Mode: emacs-lisp; lexical-binding: t; -*-

;; Copyright (C) 2001-2016  beingmeta, inc

;; Author: Ken Haase <kh@beingmeta.com>
;; Keywords: lisp framerd
;; Version: 3.9.5

;;; Commentary:

;; This package provides an emacs interaction mode for fdconsole, the
;; FramerD REPL (read-eval-print loop). It also declares various
;; indentation rules for FramerD Scheme functions.

;;; Code:

(require 'cmuscheme)

;; We do this because we don't want console windows to have infinite undo
(make-variable-buffer-local 'undo-limit)
;; The name of the FramerD scheme module for a particular buffer
(make-variable-buffer-local 'fdconsole-module)

(defvar *framerd-keywords*
  '("\\<do-choices-mt\\>" "\\<do-seq-mt\\>"
    "\\<do-choices\\>" "\\<for-choices\\>" "\\<filter-choices\\>"
    "\\<doseq\\>" "\\<dolist\\>" "\\<dotimes\\>" "\\<forseq>\\"
    "\\<lambda\\>" "\\<ambda\\>" "\\<slambda\\>" "\\<macro\\>"
    "\\<try\\>" "\\<tryif\\>"
    "\\<while\\>" "\\<until\\>" "\\<onerror>\\"
    "\\<find-frames\\>" "\\<pick\\>" "\\<reject\\>"
    "\\<div\\>" "\\<p\\>" "\\<p*\\>" "\\<form\\>"
    "\\<try-choices>\\" "\\<tryseq>\\" "\\<extdb/proc>\\"))

;; This gets #[ and #( to do block indents
(defun scheme-indent-function (indent-point state)
  (let ((normal-indent (current-column)))
    (goto-char (1+ (elt state 1)))
    (parse-partial-sexp (point) calculate-lisp-indent-last-sexp 0 t)
    (if (and (elt state 2)
	     (or (not (looking-at "\\sw\\|\\s_"))
		 (save-excursion
		   (goto-char (elt state 1))
		   (looking-at "{"))
		 (save-excursion
		   (goto-char (- (elt state 1) 1))
		   (or (looking-at "#(") (looking-at "#\\[") (looking-at "#{"))))
             ;; (not (looking-at "\\sw\\|\\s_"))
	     )
        ;; car of form doesn't seem to be a symbol
        (progn
          (if (not (> (save-excursion (forward-line 1) (point))
                      calculate-lisp-indent-last-sexp))
              (progn (goto-char calculate-lisp-indent-last-sexp)
                     (beginning-of-line)
                     (parse-partial-sexp (point)
					 calculate-lisp-indent-last-sexp 0 t)))
          ;; Indent under the list or under the first sexp on the same
          ;; line as calculate-lisp-indent-last-sexp.  Note that first
          ;; thing on that line has to be complete sexp since we are
          ;; inside the innermost containing sexp.
          (backward-prefix-chars)
          (current-column))
      (let ((function (buffer-substring (point)
					(progn (forward-sexp 1) (point))))
	    method)
	(setq method (or (get (intern-soft function) 'scheme-indent-function)
			 (get (intern-soft function) 'scheme-indent-hook)))
	(cond ((or (eq method 'defun)
		   (and (null method)
			(> (length function) 3)
			(string-match "\\`def" function)))
	       (lisp-indent-defform state indent-point))
	      ((integerp method)
	       (lisp-indent-specform method state
				     indent-point normal-indent))
	      (method
		(funcall method state indent-point normal-indent)))))))

(defun block-indenter (state indent-point normal)
  (save-excursion
    (goto-char (elt state 1))
    (+ (current-column) 2)))

;;; FRAMERD stuff
(put 'when 'scheme-indent-function 1)
(put 'unless 'scheme-indent-function 1)
(put 'tryif 'scheme-indent-function 1)
(put 'withenv 'scheme-indent-function 1)
(put 'withenv/safe 'scheme-indent-function 1)
(put 'error 'scheme-indent-function 2)
(put 'error+ 'scheme-indent-function 2)
(put 'error 'scheme-indent-function 3)
(put 'error+ 'scheme-indent-function 3)

(put 'ambda 'scheme-indent-function 1)
(put 'sambda 'scheme-indent-function 1)
(put 'slambda 'scheme-indent-function 1)

(put 'begin 'scheme-indent-function 'block-indenter)

(put 'dolist 'scheme-indent-function 1)
(put 'do-pool 'scheme-indent-function 1)
(put 'dotimes 'scheme-indent-function 1)
(put 'doseq 'scheme-indent-function 1)
(put 'dopool 'scheme-indent-function 1)
(put 'do-choices 'scheme-indent-function 1)
(put 'do-choices-mt 'scheme-indent-function 1)
(put 'do-seq-mt 'scheme-indent-function 1)
(put 'do-subsets 'scheme-indent-function 1)
(put 'for-choices 'scheme-indent-function 1)
(put 'forseq 'scheme-indent-function 1)
(put 'filter-choices 'scheme-indent-function 1)
(put 'try-choices 'scheme-indent-function 1)
(put 'tryseq 'scheme-indent-function 1)
(put 'while 'scheme-indent-function 1)
(put 'until 'scheme-indent-function 1)
(put 'with-lock 'scheme-indent-function 1)
(put 'onerror 'scheme-indent-function 1)
(put 'prog1 'scheme-indent-function 1)

(put 'index-frame 'scheme-indent-function 2)

(put 'hashtable-increment! 'scheme-indent-function 2)
(put 'hashtable-increment-existing! 'scheme-indent-function 2)
(put 'table-increment! 'scheme-indent-function 2)
(put 'table-increment-existing! 'scheme-indent-function 2)
(put 'hashtable-multiply! 'scheme-indent-function 2)
(put 'hashtable-multiply-existing! 'scheme-indent-function 2)
(put 'table-multiply! 'scheme-indent-function 2)
(put 'table-multiply-existing! 'scheme-indent-function 2)

(put 'unwind-protect 'scheme-indent-function 1)
(put 'on-errors 'scheme-indent-function 1)

(put 'printout-to 'scheme-indent-function 1)
(put 'hmac-sha1 'scheme-indent-function 1)
(put 'printout 'scheme-indent-function 'block-indenter)
(put 'lineout 'scheme-indent-function 'block-indenter)
(put 'fileout 'scheme-indent-function 1)
(put 'writeout 'scheme-indent-function 1)
(put 'writeout/type 'scheme-indent-function 2)
(put 'gp/writeout 'scheme-indent-function 2)
(put 'gp/writeout! 'scheme-indent-function 2)
(put 'gp/write! 'scheme-indent-function 3)
(put 'gp/save! 'scheme-indent-function 2)
(put 'gp/copy! 'scheme-indent-function 2)
(put 's3/op 'scheme-indent-function 3)
(put 's3/write! 'scheme-indent-function 2)
(put 's3/copy! 'scheme-indent-function 2)
(put 'ses/call 'scheme-indent-function 1)
(put 'stringout 'scheme-indent-function 'block-indenter)
(put 'glom 'scheme-indent-function 'block-indenter)
(put 'logif 'scheme-indent-function 1)
(put 'logger 'scheme-indent-function 1)
(put 'logpanic 'scheme-indent-function 1)
(put 'logalert 'scheme-indent-function 1)
(put 'logcrit 'scheme-indent-function 1)
(put 'logerr 'scheme-indent-function 1)
(put 'logerror 'scheme-indent-function 1)
(put 'logwarn 'scheme-indent-function 1)
(put 'lognotice 'scheme-indent-function 1)
(put 'loginfo 'scheme-indent-function 1)
(put 'logdebug 'scheme-indent-function 1)
(put 'logdetail 'scheme-indent-function 1)
(put 'logdeluge 'scheme-indent-function 1)
(put 'logswamp 'scheme-indent-function 1)

(put 'mongodb/find 'scheme-indent-function 2)
(put 'mongodb/update! 'scheme-indent-function 2)
(put 'mongodb/insert! 'scheme-indent-function 1)
(put 'mongodb/remove! 'scheme-indent-function 2)
(put 'mongodb/modify 'scheme-indent-function 2)
(put 'mongodb/modify! 'scheme-indent-function 2)

(put 'mongodb/cursor 'scheme-indent-function 2)
(put 'mongodb/results 'scheme-indent-function 2)
(put 'mongodb/do 'scheme-indent-function 2)

(put 'mongovec 'scheme-indent-function 'block-indenter)
(put 'mongomap 'scheme-indent-function 'block-indenter)

(put 'sig/make 'scheme-indent-function 1)
(put 'sig/check 'scheme-indent-function 2)
(put 'sig/check/ 'scheme-indent-function 3)

(put 'set-cookie! 'scheme-indent-function 1)
(put 'req/set! 'scheme-indent-function 1)
(put 'req/add! 'scheme-indent-function 1)

(put 'with-output 'scheme-indent-function 1)
(put 'with-output-to-string 'scheme-indent-function 'block-indenter)
(put 'string-subst* 'scheme-indent-function 1)

;;; XML/HTML generation
(put 'xmlout 'scheme-indent-function 'block-indenter)
(put 'xmlblock 'scheme-indent-function 2)
(put 'xmlblockn 'scheme-indent-function 2)
(put 'xmlelt 'scheme-indent-function 'block-indenter)
(put 'scripturl 'scheme-indent-function 2)
(put 'fdscripturl 'scheme-indent-function 2)

(put 'soapenvelope 'scheme-indent-function 2)

(put 'xhtml 'scheme-indent-function 'block-indenter)
(put 'span 'scheme-indent-function 1)
(put 'div 'scheme-indent-function 1)
(put 'table* 'scheme-indent-function 1)
(put 'p 'scheme-indent-function 'block-indenter)
(put 'h1 'scheme-indent-function 'block-indenter)
(put 'h2 'scheme-indent-function 'block-indenter)
(put 'h3 'scheme-indent-function 'block-indenter)
(put 'h4 'scheme-indent-function 'block-indenter)
(put 'anchor 'scheme-indent-function 1)
(put 'anchor* 'scheme-indent-function 2)
(put 'p* 'scheme-indent-function 1)
(put 'h1* 'scheme-indent-function 1)
(put 'h2* 'scheme-indent-function 1)
(put 'tr* 'scheme-indent-function 1)
(put 'th* 'scheme-indent-function 1)
(put 'td* 'scheme-indent-function 1)
(put 'ul 'scheme-indent-function 1)
(put 'li 'scheme-indent-function 1)
(put 'ul* 'scheme-indent-function 1)
(put 'li* 'scheme-indent-function 1)
(put 'vistoggle 'scheme-indent-function 1)
(put 'form 'scheme-indent-function 1)

(put 'find-frames 'scheme-indent-function 1)
(put 'frame-create 'scheme-indent-function 1)

(put '%watch 'scheme-indent-function 1)
(put 'deluge%watch 'scheme-indent-function 1)
(put 'detail%watch 'scheme-indent-function 1)
(put 'debug%watch 'scheme-indent-function 1)
(put 'info%watch 'scheme-indent-function 1)
(put 'notice%watch 'scheme-indent-function 1)
(put 'warn%watch 'scheme-indent-function 1)
(put 'always%watch 'scheme-indent-function 1)
(put 'saveoutput 'scheme-indent-function 1)

(put 'extdb/proc 'scheme-indent-function 1)

;;;; Evaluating expressions in modules

(defvar fdconsole-module)
(defconst in-module-regexp
  "(in-module +'\\(\\(\\w\\|[/$.-_]\\)+\\)")

(defun fdconsole-get-module-name ()
  "Returns the module name specified in the current buffer"
  (if (and (boundp 'fdconsole-module) fdconsole-module) fdconsole-module
    (save-excursion
      (goto-char (point-min))
      (let* ((pos (search-forward-regexp in-module-regexp (point-max) t))
	     (name
	      (if pos
		  (buffer-substring (match-beginning 1) (match-end 1))
		"")))
;	(if pos
;	    (message "Search found module name %s at %d" name pos)
;	  (message "Search failed to find module name"))
	(if pos (setq fdconsole-module name))
	name))))
(defun fdconsole-process () (scheme-proc))

(defun fdconsole-send-region (start end)
  (let ((module (fdconsole-get-module-name))
	(process (fdconsole-process)))
    (if (not (equal module ""))
	(message "Sending %d characters into the %s module"
		 (- end start) module))
    (if (not (equal module ""))
	(comint-send-string
	 process (format "(within-module '%s\n" module)))
    (comint-send-region process start end)
    (if (not (equal module ""))
	(comint-send-string process ")\n")
      (comint-send-string process "\n"))))

(defun fdconsole-send-definition ()
  "Send the current definition to the inferior Scheme process."
  (interactive)
  (save-excursion
   (end-of-defun)
   (let ((end (point)))
     (beginning-of-defun)
     (fdconsole-send-region (point) end))))

(defun scheme-send-last-sexp ()
  "Send the previous sexp to the inferior Scheme process."
  (interactive)
  (fdconsole-send-region (save-excursion (backward-sexp) (point)) (point)))

(defun fdconsole-sender ()
  (interactive)
  (if mark-active (fdconsole-send-region (region-beginning) (region-end))
    (fdconsole-send-definition)))

(defun split-command-line (string)
  (let ((where (string-match "[ \t]" string)))
    (cond ((null where) (list string))
	  ((not (= where 0))
	   (cons (substring string 0 where)
		 (split-command-line (substring string (+ 1 where)
						(length string)))))
	  (t (let ((pos (string-match "[^ \t]" string)))
	       (if (null pos)
		   nil
		 (split-command-line (substring string pos
						(length string)))))))))

;;; Running an fdconsole

(defvar fdconsole-program "fdconsole")

(autoload 'comint-check-proc "comint")

(defun fdconsole (cmd)
  "Run an inferior FramerD scheme process, input and output via buffer *fdconsole*.
With an arguments, prompts for a command and arguments to use.
If there is a process already running in `*fdconsole*', switch to that buffer.
Runs the hooks `inferior-scheme-mode-hook' \(after the `comint-mode-hook' is
run). \(Type \\[describe-mode] in the process buffer for a list of commands.)"
  (interactive
   (list (if current-prefix-arg
	     (read-string "Run fdconsole: " fdconsole-program)
	   fdconsole-program)))
  (let ((bufname (or (and scheme-buffer
			  (get-buffer-window scheme-buffer)
			  scheme-buffer)
		     "*fdconsole*"))
	(comint-arg "fdconsole"))
    (if (equal major-mode (intern "inferior-scheme-mode"))
	(progn (setq bufname (buffer-name (current-buffer)))
	       (setq comint-arg bufname)
	       (if (equal (substring comint-arg 0 1) "*")
		   (setq comint-arg (substring comint-arg 1)))
	       (if (equal (substring comint-arg -1) "*")
		   (setq comint-arg (substring comint-arg 0 -1)))))
    (if (not (comint-check-proc bufname))
	(let ((cmdlist (split-command-line cmd)))
	  (set-buffer (apply 'make-comint comint-arg (car cmdlist)
			     nil (cdr cmdlist)))
	  (inferior-scheme-mode)))
    (setq scheme-program-name cmd)
    (setq scheme-buffer bufname)
    (pop-to-buffer bufname)))

;;; Defining a mode hook to define fdconsole-sender

(defun fdconsole-scheme-mode-hook ()
  (interactive)
  (local-set-key "\e\C-m" 'fdconsole-sender)
  (setq undo-limit 32)
  (font-lock-add-keywords 'scheme-mode *framerd-keywords*))
(add-hook 'scheme-mode-hook 'fdconsole-scheme-mode-hook)


(provide 'fdconsole)
;;; fdconsole.el ends here

